zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java

```diff
@@ -1350,6 +1350,15 @@ public void run() {
                     ServerMetrics.getMetrics().LOOKING_COUNT.add(1);
 
                     if (Boolean.getBoolean("readonlymode.enabled")) {
+                        if (!zkDb.isInitialized()) {
```

Comment:
If there is issue before finishing syncing with leader, the proper behavior probably is exit instead of blocking here. Otherwise, it won't move forward even after the network issue recovered, let me know if I missed anything.

Reply:
Thank you for comments.
Let's take Follower as example.
In my understanding, FollowerZooKeeperServer is shutdown and clear QuorumPeer's `follower` as null when there is issue during syncing with leader. This is as same as what you described. But the corresponding QuorumPeer is still alive and it will fall into ServerState.LOOKING. In the LOOKING state, it will start a temporary ReadOnlyZooKeeperServer if `readonlymode.enabled` is set. That's the point that this PR want to resolve.
If `readonlymode.enabled` is not set, I think it wont't move forward and look for leader.
So I missed anything ? Thank you!

Reply:
@yfxhust sorry for the lately reply, if I remember correctly, the zk database will be loaded when we started the ZK process for the first time, and might be cleared when have exception following/observing leader, and only reloaded when start to look for leader triggered with the following code:

    setCurrentVote(makeLEStrategy().lookForLeader());

In which case, it may cause it blocking here, let me know if this is correct or not.

Reply:
@lvfangmin Thank you for reply. Yes. I think you are right. But my patch tries to address the following scenarios. Maybe you can help to figure the mistake.

1. At time T0, Observer O1 loses all the network connection. The network broken lasts for a long time(Maybe we can use iptables drop to simulate this network broken). So observer O1 fall into LOOKING state and serves as readonly server. Everything is OK here.

2. At time T1, Observer O1 network broken is recovered and it is connected to Leader again. But because the quorum cluster has too much new txn and observer falls behind too much, O1 choose to use deserializeSnapshot() to sync ZKDatabase with leader. deserializeSnapshot() firstly uses clear() to clear ZKDatabase and then uses SerializeUtils.deserializeSnapshot() to recover dataTree. The time window between clear() and SerializeUtils.deserializeSnapshot() is dangerous because the dataTree may be null during this window. I think this dangerous window is not protected internally and can be catched by Zookeeper client.

3. At time T2, Observer O1 finished the sync with leader. It quits from the blocking setCurrentVote(makeLEStrategy().lookForLeader()) and shutdown readonly server.

So please let me know if I missed something in above inference.
If my inference is true, my patch try to fix the dangerous window in step2(at time T1).

Thank you!